drivers/interrupt_controller/intc_gpio_stm32.c

```diff
@@ -0,0 +1,317 @@
+};
+
+static struct stm32_intc_gpio_data intc_gpio_data;
+
+/**
+ * @returns the LL_<PPP>_EXTI_LINE_xxx define that corresponds to specified @p linenum
+ * This value can be used with the LL EXTI source configuration functions.
+ */
+static inline uint32_t stm32_exti_linenum_to_src_cfg_line(gpio_pin_t linenum)
+{
+#if defined(CONFIG_SOC_SERIES_STM32L0X) || \
+	defined(CONFIG_SOC_SERIES_STM32F0X)
+	return ((linenum % 4 * 4) << 16) | (linenum / 4);
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32g0_exti)
+	return ((linenum & 0x3) << (16 + 3)) | (linenum >> 2);
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7rs_exti)
+	/* Gives the LL_SBS_EXTI_LINEn corresponding to the line number */
+	return (((linenum % 4 * 4) << LL_SBS_REGISTER_PINPOS_SHFT) | (linenum / 4));
+#else
+	return (0xF << ((linenum % 4 * 4) + 16)) | (linenum / 4);
+#endif
+}
+
+/**
+ * @returns EXTI line number for LL_EXTI_LINE_n define
+ */
+static inline gpio_pin_t ll_exti_line_to_linenum(stm32_gpio_irq_line_t line)
+{
+	return LOG2(line);
+}
+
+/**
+ * @brief EXTI ISR handler
+ *
+ * Check EXTI lines in exti_range for pending interrupts
+ *
+ * @param exti_range Pointer to a exti_range structure
+ */
+static void stm32_intc_gpio_isr(const void *exti_range)
+{
+	struct stm32_intc_gpio_data *data = &intc_gpio_data;
+	const struct stm32_exti_range *range = exti_range;
+	stm32_gpio_irq_line_t line;
+	uint32_t line_num;
+
+	/* see which bits are set */
+	for (uint8_t i = 0; i <= range->len; i++) {
+		line_num = range->start + i;
+		line = exti_linenum_to_ll_exti_line(line_num);
+
+		/* check if interrupt is pending */
+		if (stm32_exti_is_pending(line) != 0) {
+			/* clear pending interrupt */
+			stm32_exti_clear_pending(line);
+
+			/* run callback only if one is registered */
+			if (!data->cb[line_num].cb) {
+				continue;
+			}
+
+			/* `line` can be passed as-is because LL_EXTI_LINE_n is (1 << n) */
+			data->cb[line_num].cb(line, data->cb[line_num].data);
+		}
+	}
+}
+
+static void stm32_fill_irq_table(int8_t start, int8_t len, int32_t irqn)
+{
+	for (int i = 0; i < len; i++) {
+		exti_irq_table[start + i] = irqn;
+	}
+}
+
+/* This macro:
+ * - populates line_range_x from line_range dt property
+ * - fill exti_irq_table through stm32_fill_irq_table()
+ * - calls IRQ_CONNECT for each interrupt and matching line_range
+ */
+#define STM32_EXTI_INIT_LINE_RANGE(node_id, interrupts, idx)			\
+	static const struct stm32_exti_range line_range_##idx = {		\
+		DT_PROP_BY_IDX(node_id, line_ranges, UTIL_X2(idx)),		\
+		DT_PROP_BY_IDX(node_id, line_ranges, UTIL_INC(UTIL_X2(idx)))	\
+	};									\
+	stm32_fill_irq_table(line_range_##idx.start,				\
+			     line_range_##idx.len,				\
+			     DT_IRQ_BY_IDX(node_id, idx, irq));			\
+	IRQ_CONNECT(DT_IRQ_BY_IDX(node_id, idx, irq),				\
+		DT_IRQ_BY_IDX(node_id, idx, priority),				\
+		stm32_intc_gpio_isr, &line_range_##idx, 0);
+
+/**
+ * @brief Initializes the EXTI GPIO interrupt controller driver
+ */
+static int stm32_intc_gpio_init(void)
```

Comment:
Nit: I would prefer the following name
```suggestion
static int stm32_exti_gpio_intc_init(void)
```